### PR TITLE
Provide API to preserve source range and callstack information during graph rewrite

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58492 [Pytorch] Update FuseLinear to map source range information
* **#58300 [Pytorch] Provide API to preserve source range and callstack information during graph rewrite**

Summary:
Current state: During graph rewriting that can fuse nodes or add nodes
result in new nodes without debug information that was available in
original node. Thus we lose this information during graph rewrite.

This PR changes graph rewriting API to let user specify how the values
in the replacement pattern map to values in the pattern to be matched.
Then the graph rewriting will copy source range and inlined callstack
from the matched nodes onto the nodes being inserted.

Test Plan:
python test/test_jit.py
TestJit.test_pattern_based_rewrite_with_source_range_preserved

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28512465](https://our.internmc.facebook.com/intern/diff/D28512465)

### CODE REVIEW

File: torch/csrc/jit/passes/subgraph_rewrite.cpp [hunk 1]
```diff
@@ -17,13 +43,16 @@ graph(%x, %w, %b):
       R"IR(
 graph(%x, %w, %b):
   %r = aten::convrelu(%x, %w, %b)
-  return (%r))IR");
+  return (%r))IR",
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/subgraph_rewrite.cpp:46]: Do we use the default patterns anywhere btw? I added them in the beginning, but I suspect we register patterns explicitly in all use cases - if that's the case we could probably nuke this.

File: torch/csrc/jit/passes/subgraph_rewrite.cpp [hunk 2]
```diff
@@ -56,7 +85,26 @@ void SubgraphRewriter::rewriteSinglePatternOnGraph(
   parseIR(pattern.pattern, &pattern_graph, vmap);
 
   Graph replacement_graph;
-  parseIR(pattern.replacement, &replacement_graph);
+  std::unordered_map<std::string, Value*> vmap_replacement;
+  parseIR(pattern.replacement, &replacement_graph, vmap_replacement);
+
+  // First construct map of Node*-to-Node*
+  // This maps Nodes in replacement graph to nodes in pattern graph
+  // given the value_name_map, which maps value names from repalcement
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/subgraph_rewrite.cpp:93]: Typo: repalcment

**Review by ZolotukhinM — APPROVED** [review]
LGTM!

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28512465).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28512465).

**Comment by kimishpatel** [comment]
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28512465).

**Pull request closed without merging**